### PR TITLE
Bring back firejail support

### PR DESCRIPTION
**Pull request by azubieta**

### CODE REVIEW

**Comment by probonopd** [comment]
Thanks @azubieta - can you give the build artifact from Travis CI a try and see whether it works with and without Firejail? Thanks.

File: src/main.c [hunk 1]
```diff
@@ -150,12 +150,76 @@ bool is_appimage(char *path, gboolean verbose)
...
+    g_key_file_set_value(key_file_structure, firejail_profile_group, G_KEY_FILE_DESKTOP_KEY_EXEC,
+                         firejail_profile_exec);
+    g_key_file_set_value(key_file_structure, firejail_profile_group, G_KEY_FILE_DESKTOP_KEY_TRY_EXEC,
+                         firejail_tryexec);
+    g_key_file_set_value(key_file_structure, G_KEY_FILE_DESKTOP_GROUP, "Actions", "FirejailProfile;");
+}
+
+void save_desktop_entry(GKeyFile* key_file_structure, const char* desktop_file_path) {
+    gboolean success = g_key_file_save_to_file(key_file_structure, desktop_file_path, NULL);
+    if (!success) {
```
Comment (TheAssassin) [inline src/main.c]: Empty line missing

File: src/main.c [hunk 2]
```diff
@@ -150,12 +150,76 @@ bool is_appimage(char *path, gboolean verbose)
...
+        // Don't remove the dbrackets or the macro will segfault
```
Comment (TheAssassin) [inline src/main.c]: Is `dbrackets` a typo or does it have a special meaning?

File: src/main.c [hunk 3]
```diff
@@ -150,12 +150,76 @@ bool is_appimage(char *path, gboolean verbose)
...
+        THREADSAFE_G_PRINT("Failed to load the deployed desktop entry, '%s'\n", "a");
+    }
+
+
+    return key_file_structure;
+}
+
+void setup_firejail_on_desktop_entry(GKeyFile* key_file_structure) {
+    char* firejail_exec;
+    char* oldExecValue = g_key_file_get_value(key_file_structure,
```
Comment (TheAssassin) [inline src/main.c]: This is not freed.

File: src/main.c [hunk 4]
```diff
@@ -150,12 +150,76 @@ bool is_appimage(char *path, gboolean verbose)
...
+        // Don't remove the dbrackets or the macro will segfault
+        THREADSAFE_G_PRINT("Failed to load the deployed desktop entry, '%s'\n", "a");
+    }
+
+
+    return key_file_structure;
+}
+
+void setup_firejail_on_desktop_entry(GKeyFile* key_file_structure) {
+    char* firejail_exec;
```
Comment (TheAssassin) [inline src/main.c]: There is no reason to declare the variable here and not next to its initialization.

File: src/main.c [hunk 5]
```diff
@@ -150,12 +150,76 @@ bool is_appimage(char *path, gboolean verbose)
...
+    GKeyFile* key_file_structure = g_key_file_new();
+    gboolean success = g_key_file_load_from_file(key_file_structure, desktop_file_path,
+                                                 G_KEY_FILE_KEEP_COMMENTS | G_KEY_FILE_KEEP_TRANSLATIONS, NULL);
+
+    if (!success) {
+        // Don't remove the dbrackets or the macro will segfault
+        THREADSAFE_G_PRINT("Failed to load the deployed desktop entry, '%s'\n", "a");
+    }
+
+
```
Comment (TheAssassin) [inline src/main.c]: Too many blank lines.

File: src/main.c [hunk 6]
```diff
@@ -150,12 +150,76 @@ bool is_appimage(char *path, gboolean verbose)
...
+    }
+}
+
+void* thread_appimage_register_in_system(void* arguments) {
+    struct arg_struct* args = arguments;
+    if (is_appimage(args->path, args->verbose) && !appimage_is_registered_in_system(args->path)) {
+        int failed = appimage_register_in_system(args->path, args->verbose);
+        if (!failed) {
+            enable_firejail_if_available(args->path);
+            update_desktop_set_dirty();
```
[the remark this answers is not in the document]
Comment (azubieta) [reply src/main.c]: No clue!

File: src/main.c (hunk 2 above)
Comment (azubieta) [reply src/main.c]: typo, fixed

File: src/main.c (hunk 3 above)
Comment (azubieta) [reply src/main.c]: fixed

File: src/main.c (hunk 1 above)
Comment (azubieta) [reply src/main.c]: fiexed

File: src/main.c [hunk 7]
```diff
@@ -150,12 +150,76 @@ bool is_appimage(char *path, gboolean verbose)
...
+            g_key_file_unref(key_file_structure);
+        }
+    }
+}
+
+void* thread_appimage_register_in_system(void* arguments) {
+    struct arg_struct* args = arguments;
+    if (is_appimage(args->path, args->verbose) && !appimage_is_registered_in_system(args->path)) {
+        int failed = appimage_register_in_system(args->path, args->verbose);
+        if (!failed) {
```
[the remark this answers is not in the document]
Comment (azubieta) [reply src/main.c]: fixed

File: src/main.c (hunk 4 above)
Comment (azubieta) [reply src/main.c]: fixed

File: src/main.c (hunk 5 above)
Comment (azubieta) [reply src/main.c]: fiexed

File: src/main.c [hunk 8]
```diff
@@ -150,12 +150,76 @@ bool is_appimage(char *path, gboolean verbose)
...
+        // Don't remove the dbrackets or the macro will segfault
+        THREADSAFE_G_PRINT("Failed to save the deployed desktop entry\n");
+    }
+}
+
+
+void enable_firejail_if_available(const char* appimage_path) {
+    /* If firejail is on the $PATH, then use it to run AppImages */
+    if (g_find_program_in_path("firejail")) {
+        const char* desktop_file_path = appimage_registered_desktop_file_path(appimage_path, NULL, false);
```
[the remark this answers is not in the document]
Comment (azubieta) [reply src/main.c]: fiexed

**Comment by azubieta** [comment]
@probonopd I just ran `appimaged` using `valgrind` it has memory leaks everywhere. Maintaining this code while having other superior (attending to memory management and maintainability) implementations don't seem like a good idea to me.

I might be a good time to freeze it after this commit and move on `appimagelauncherd` as the default integration tool.  It is only missing your specific requirements, right @TheAssassin ?

**Comment by probonopd** [comment]
`appimaged` is a main component of how the AppImage system needs to work (for me, anyways) and I am not going to accept it becoming unusable. I will not be going away.

If we can improve `appimaged` by borrowing some code from `appimagelauncherd` then fine, but I don't want _yet another_ new repo, new name, new instructions and so on. Having moved `appimaged` out of the AppImageKit repo was already a hassle that I didn't really like. For me, `appimaged` (apart from the performance and stability issues) does exactly what it needs to do, and I don't want or need something else.

File: src/main.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (TheAssassin) [reply src/main.c]: Why do you use it then?

[the remark this answers is not in the document]
Comment (azubieta) [reply src/main.c]: After reading the code I found that it sets a flag to tell whether there have been changes in the watched directories that might require to update the desktop menu.

The code was already there, I just move it up.

**Comment by azubieta** [comment]
@probono it took the freedom to move your comment about your intentions with `appimaged` to a new issue. Just to keep focused here in bringing back the `firejail` support.

**Pull request merged**